src/main/java/org/broadinstitute/hellbender/utils/Utils.java

```diff
@@ -1057,6 +1057,114 @@ public static int lastIndexOf(final byte[] reference, final byte[] query) {
         return -1;
     }
 
+    /**
+     * Find the last one-off occurrence of the query sequence in the reference sequence
+     *
+     * Returns the index of the last occurrence or -1 if the query sequence is not found
+     *
+     * @param reference the reference sequence
+     * @param query the query sequence
+     */
+    public static int lastIndexOfAtMostOneMismatch(final byte[] reference, final byte[] query) {
+        int queryLength = query.length;
+
+        // start search from the last possible matching position and search to the left
+        for (int refIndex = reference.length - queryLength; refIndex >= 0; refIndex--) {
+            int mismatchCount = 0;
+            for (int queryIndex = 0; queryIndex < queryLength && mismatchCount < 2; queryIndex++) {
+                if (reference[refIndex+queryIndex] != query[queryIndex]) {
+                    mismatchCount++;
+                }
+            }
+
+            if (mismatchCount < 2) {
+                return refIndex;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Finds the location of one indel in the query sequence in relation to the reference sequence
+     *
+     * Returns the index and size of the indel (as a 2 element int array) or -1 and 0 if an indel less than 4 bases is not found
+     *
+     * @param reference the reference sequence
+     * @param query the query sequence
+     * @param allowedLengthOfIndel the maximum size of the indel before 2 mismatches becomes less penalized
+     */
+    public static int[] atMostOneIndel(final byte[] reference, final byte[] query, int allowedLengthOfIndel){
+        int referenceLength = reference.length;
+        int queryLength = query.length;
+
+        int indelEnd = -1;
+        int[] indelStartAndSize = {-1,0};
+
+        if(queryLength < referenceLength){
+            int lengthOfIndel = referenceLength - queryLength;
+            if(lengthOfIndel > allowedLengthOfIndel){
+                return indelStartAndSize;
+            }
+
+            //traverse backwards until you hit mismatch/end of indel
+            for(int i = queryLength - 1; i >= 0; i--){
```

Comment:
You could reduce the amount of nesting by rewriting this as:

* Go backwards from the query end until you find a mismatch.  This tells you where the indel must start.  This could be implemented as a `for` loop with a `break`, or a `while` loop, or something like (my vote, but I will defer to anybody on the engine team): `final int indelEnd = IntStream.range(0,queryLength).map(n -> queryLength - n).filter(n -> query[n] != reference[n]).findFirst().orElse(-1)`.
* Go forward with something similar until you find where the indel must start.
* Figure out the indel in the middle.